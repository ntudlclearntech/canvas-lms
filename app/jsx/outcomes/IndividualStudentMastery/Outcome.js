/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import I18n from 'i18n!outcomes'
import View from '@instructure/ui-layout/lib/components/View'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import ToggleDetails from '@instructure/ui-toggle-details/lib/components/ToggleDetails'
import List, { ListItem } from '@instructure/ui-elements/lib/components/List'
import Pill from '@instructure/ui-elements/lib/components/Pill'
import Text from '@instructure/ui-elements/lib/components/Text'
import IconArrowOpenDown from '@instructure/ui-icons/lib/Solid/IconArrowOpenDown'
import IconArrowOpenRight from '@instructure/ui-icons/lib/Solid/IconArrowOpenRight'
import IconOutcomes from '@instructure/ui-icons/lib/Line/IconOutcomes'
import AssignmentResult from './AssignmentResult'
import * as shapes from './shapes'

const spacyIcon = (expanded) => () => {
  const Icon = expanded ? IconArrowOpenDown : IconArrowOpenRight
  return (
    <View padding="0 0 0 small"><Icon /></View>
  )
}

export default class Outcome extends React.Component {
  static propTypes = {
    outcome: shapes.outcomeShape.isRequired
  }

  constructor () {
    super()
    this.handleToggle = this.handleToggle.bind(this)
    this.state = { expanded: false }
  }

  contract () {
    this.setState({ expanded: false })
  }

  expand () {
    this.setState({ expanded: true })
  }

  handleToggle (_event, expanded) {
    this.setState({ expanded })
  }

  renderHeader () {
    const { outcome } = this.props
    const { mastered, results, title } = outcome
    const numAlignments = results.length

    return (
      <Flex direction="row" justifyItems='space-between' padding="small x-small">
        <FlexItem>
          <Flex direction="column">
            <FlexItem>
              <Text size="medium">
                <Flex>
                  <FlexItem><IconOutcomes /></FlexItem>
                  <FlexItem padding="0 x-small">{ title }</FlexItem>
                </Flex>
              </Text>
            </FlexItem>
            <FlexItem><Text size="small">{ I18n.t({
              zero: 'No alignments',
              one: '%{count} alignment',
              other: '%{count} alignments'
            }, { count: I18n.n(numAlignments) }) }</Text></FlexItem>
          </Flex>
        </FlexItem>
        <FlexItem>
        {
          mastered ? <Pill text={I18n.t('Mastered')} variant="success" /> : <Pill text={I18n.t('Not mastered')} />
        }
        </FlexItem>
      </Flex>
    )
  }

  render () {
    const { outcome } = this.props
    return (
      <ToggleDetails
        divider="dashed"
        icon={spacyIcon(false)}
        iconExpanded={spacyIcon(true)}
        summary={this.renderHeader()}
        expanded={this.state.expanded}
        onToggle={this.handleToggle}
        fluidWidth
      >
        <View as="div" borderWidth="small 0 0 0">
          <List variant="unstyled" delimiter="dashed">
          {
            outcome.results.map((result) => (
              <ListItem key={result.id}><AssignmentResult result={result} outcome={outcome} /></ListItem>
            ))
          }
          </List>
        </View>
      </ToggleDetails>
    )
  }
}
