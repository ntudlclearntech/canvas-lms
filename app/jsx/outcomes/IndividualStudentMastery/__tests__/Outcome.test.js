/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render, shallow } from 'enzyme'
import Outcome from '../Outcome'

const defaultProps = (props = {}) => (
  Object.assign({
    outcome: {
      id: 1,
      mastered: false,
      ratings: [
        { description: 'My first rating' },
        { description: 'My second rating' }
      ],
      results: [
        {
          id: 1,
          score: 1,
          percent: 0.1,
          assignment: {
            id: 1,
            html_url: 'http://foo',
            name: 'My assignment',
            submission_types: 'online_quiz'
          }
        }
      ],
      title: 'My outcome'
    }
  }, props)
)

it('renders the Outcome component', () => {
  const wrapper = shallow(<Outcome {...defaultProps()}/>)
  expect(wrapper.debug()).toMatchSnapshot()
})

describe('header', () => {
  it('includes the outcome name', () => {
    const wrapper = shallow(<Outcome {...defaultProps()}/>)
    const header = wrapper.find('ToggleDetails')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('My outcome')
  })

  it('includes mastery when mastered', () => {
    const props = defaultProps()
    props.outcome.mastered = true
    const wrapper = shallow(<Outcome {...props} />)
    const header = wrapper.find('ToggleDetails')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('Mastered')
  })

  it('includes non-mastery when not mastered', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />)
    const header = wrapper.find('ToggleDetails')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('Not mastered')
  })

  it('shows correct number of alignments', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />)
    const header = wrapper.find('ToggleDetails')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('1 alignment')
  })
})

it('includes the individual results', () => {
  const wrapper = shallow(<Outcome {...defaultProps()} />)
  expect(wrapper.find('AssignmentResult')).toHaveLength(1)
})

it('defaults to unexpanded', () => {
  const wrapper = shallow(<Outcome {...defaultProps()} />)
  expect(wrapper.state('expanded')).toBe(false)
})

describe('expand()', () => {
  it('expands when called', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />)
    wrapper.instance().expand()
    expect(wrapper.state('expanded')).toBe(true)
  })
})

describe('contract()', () => {
  it('contracts when called', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />).setState({ expanded: true })
    wrapper.instance().contract()
    expect(wrapper.state('expanded')).toBe(false)
  })
})

describe('handleToggle()', () => {
  it('expands when called with true', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />)
    wrapper.instance().handleToggle(null, true)
    expect(wrapper.state('expanded')).toBe(true)
  })

  it('contracts when called with false', () => {
    const wrapper = shallow(<Outcome {...defaultProps()} />).setState({ expanded: true })
    wrapper.instance().handleToggle(null, false)
    expect(wrapper.state('expanded')).toBe(false)
  })
})
