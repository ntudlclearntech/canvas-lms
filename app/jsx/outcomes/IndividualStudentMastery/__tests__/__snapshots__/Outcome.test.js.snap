// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly expanded 1`] = `
"<ToggleDetails divider=\\"dashed\\" icon={[Function]} iconExpanded={[Function]} summary={{...}} expanded={true} onToggle={[Function]} fluidWidth={true} variant=\\"default\\" size=\\"medium\\" iconPosition=\\"start\\" defaultExpanded={false}>
  <View as=\\"div\\" borderWidth=\\"small 0 0 0\\" display=\\"auto\\">
    <List variant=\\"unstyled\\" delimiter=\\"dashed\\" as=\\"ul\\" margin=\\"none\\" size=\\"medium\\">
      <ListItem>
        <AssignmentResult result={{...}} outcome={{...}} />
      </ListItem>
    </List>
  </View>
</ToggleDetails>"
`;

exports[`renders correctly expanded with no results 1`] = `
"<ToggleDetails divider=\\"dashed\\" icon={[Function]} iconExpanded={[Function]} summary={{...}} expanded={true} onToggle={[Function]} fluidWidth={true} variant=\\"default\\" size=\\"medium\\" iconPosition=\\"start\\" defaultExpanded={false}>
  <View as=\\"div\\" padding=\\"small\\" borderWidth=\\"small 0 0 0\\" display=\\"auto\\">
    <Text as=\\"span\\" size=\\"medium\\" letterSpacing=\\"normal\\">
      No alignments are available for this outcome.
    </Text>
  </View>
</ToggleDetails>"
`;

exports[`renders the Outcome component 1`] = `
"<ToggleDetails divider=\\"dashed\\" icon={[Function]} iconExpanded={[Function]} summary={{...}} expanded={false} onToggle={[Function]} fluidWidth={true} variant=\\"default\\" size=\\"medium\\" iconPosition=\\"start\\" defaultExpanded={false}>
  <View as=\\"div\\" borderWidth=\\"small 0 0 0\\" display=\\"auto\\">
    <List variant=\\"unstyled\\" delimiter=\\"dashed\\" as=\\"ul\\" margin=\\"none\\" size=\\"medium\\">
      <ListItem>
        <AssignmentResult result={{...}} outcome={{...}} />
      </ListItem>
    </List>
  </View>
</ToggleDetails>"
`;
